// mdb-angular-ui-kit/bundles/mdb-angular-ui-kit-popover.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/cdk/overlay'), require('@angular/cdk/portal'), require('@angular/animations'), require('rxjs'), require('rxjs/operators'), require('@angular/cdk/coercion'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('mdb-angular-ui-kit/popover', ['exports', '@angular/core', '@angular/cdk/overlay', '@angular/cdk/portal', '@angular/animations', 'rxjs', 'rxjs/operators', '@angular/cdk/coercion', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['mdb-angular-ui-kit'] = global['mdb-angular-ui-kit'] || {}, global['mdb-angular-ui-kit'].popover = {}), global.ng.core, global.ng.cdk.overlay, global.ng.cdk.portal, global.ng.animations, global.rxjs, global.rxjs.operators, global.ng.cdk.coercion, global.ng.common));
}(this, (function (exports, core, overlay, portal, animations, rxjs, operators, coercion, common) { 'use strict';

    var MdbPopoverComponent = /** @class */ (function () {
        function MdbPopoverComponent(_cdRef) {
            this._cdRef = _cdRef;
            this._hidden = new rxjs.Subject();
            this.animationState = 'hidden';
        }
        Object.defineProperty(MdbPopoverComponent.prototype, "isContentTemplate", {
            get: function () {
                return this.content instanceof core.TemplateRef;
            },
            enumerable: false,
            configurable: true
        });
        MdbPopoverComponent.prototype.markForCheck = function () {
            this._cdRef.markForCheck();
        };
        MdbPopoverComponent.prototype.onAnimationEnd = function (event) {
            if (event.toState === 'hidden') {
                this._hidden.next();
            }
        };
        return MdbPopoverComponent;
    }());
    MdbPopoverComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'mdb-popover',
                    template: "<div\n  class=\"popover\"\n  [@fade]=\"animationState\"\n  (@fade.done)=\"onAnimationEnd($event)\"\n  [@.disabled]=\"!animation\"\n>\n  <div *ngIf=\"title\" class=\"popover-header\">\n    {{ title }}\n  </div>\n  <div *ngIf=\"isContentTemplate\" class=\"popover-body\">\n    <ng-container [ngTemplateOutlet]=\"content\"></ng-container>\n  </div>\n  <div *ngIf=\"!isContentTemplate\" class=\"popover-body\">\n    {{ content }}\n  </div>\n</div>\n",
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                    animations: [
                        animations.trigger('fade', [
                            animations.state('visible', animations.style({ opacity: 1 })),
                            animations.state('hidden', animations.style({ opacity: 0 })),
                            animations.transition('visible <=> hidden', animations.animate('150ms linear')),
                            animations.transition(':enter', animations.animate('150ms linear')),
                        ]),
                    ]
                },] }
    ];
    MdbPopoverComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef }
    ]; };
    MdbPopoverComponent.propDecorators = {
        title: [{ type: core.Input }],
        content: [{ type: core.Input }],
        animation: [{ type: core.Input }]
    };

    // eslint-disable-next-line @angular-eslint/component-class-suffix
    var MdbPopoverDirective = /** @class */ (function () {
        function MdbPopoverDirective(_overlay, _overlayPositionBuilder, _elementRef) {
            this._overlay = _overlay;
            this._overlayPositionBuilder = _overlayPositionBuilder;
            this._elementRef = _elementRef;
            this.mdbPopover = '';
            this.mdbPopoverTitle = '';
            this._popoverDisabled = false;
            this.placement = 'top';
            this._animation = false;
            this.trigger = 'click';
            this.delayShow = 0;
            this.delayHide = 0;
            this.offset = 4;
            this.popoverShow = new core.EventEmitter();
            this.popoverShown = new core.EventEmitter();
            this.popoverHide = new core.EventEmitter();
            this.popoverHidden = new core.EventEmitter();
            this._open = false;
            this._showTimeout = 0;
            this._hideTimeout = 0;
            this._destroy$ = new rxjs.Subject();
        }
        Object.defineProperty(MdbPopoverDirective.prototype, "popoverDisabled", {
            get: function () {
                return this._popoverDisabled;
            },
            set: function (value) {
                this._popoverDisabled = coercion.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(MdbPopoverDirective.prototype, "animation", {
            get: function () {
                return this._animation;
            },
            set: function (value) {
                this._animation = coercion.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        MdbPopoverDirective.prototype.ngOnInit = function () {
            if (this.popoverDisabled) {
                return;
            }
            this._bindTriggerEvents();
        };
        MdbPopoverDirective.prototype.ngOnDestroy = function () {
            if (this._open) {
                this.hide();
            }
            this._destroy$.next();
            this._destroy$.complete();
        };
        MdbPopoverDirective.prototype._bindTriggerEvents = function () {
            var _this = this;
            var triggers = this.trigger.split(' ');
            triggers.forEach(function (trigger) {
                if (trigger === 'click') {
                    rxjs.fromEvent(_this._elementRef.nativeElement, trigger)
                        .pipe(operators.takeUntil(_this._destroy$))
                        .subscribe(function () { return _this.toggle(); });
                }
                else if (trigger !== 'manual') {
                    var evIn = trigger === 'hover' ? 'mouseenter' : 'focusin';
                    var evOut = trigger === 'hover' ? 'mouseleave' : 'focusout';
                    rxjs.fromEvent(_this._elementRef.nativeElement, evIn)
                        .pipe(operators.takeUntil(_this._destroy$))
                        .subscribe(function () { return _this.show(); });
                    rxjs.fromEvent(_this._elementRef.nativeElement, evOut)
                        .pipe(operators.takeUntil(_this._destroy$))
                        .subscribe(function () { return _this.hide(); });
                }
            });
        };
        MdbPopoverDirective.prototype._createOverlayConfig = function () {
            var positionStrategy = this._overlayPositionBuilder
                .flexibleConnectedTo(this._elementRef)
                .withPositions(this._getPosition());
            var overlayConfig = new overlay.OverlayConfig({
                hasBackdrop: false,
                scrollStrategy: this._overlay.scrollStrategies.reposition(),
                positionStrategy: positionStrategy,
            });
            return overlayConfig;
        };
        MdbPopoverDirective.prototype._createOverlay = function () {
            this._overlayRef = this._overlay.create(this._createOverlayConfig());
        };
        MdbPopoverDirective.prototype._getPosition = function () {
            var position;
            var positionTop = {
                originX: 'center',
                originY: 'top',
                overlayX: 'center',
                overlayY: 'bottom',
                offsetY: -this.offset,
            };
            var positionBottom = {
                originX: 'center',
                originY: 'bottom',
                overlayX: 'center',
                overlayY: 'top',
                offsetY: this.offset,
            };
            var positionRight = {
                originX: 'end',
                originY: 'center',
                overlayX: 'start',
                overlayY: 'center',
                offsetX: this.offset,
            };
            var positionLeft = {
                originX: 'start',
                originY: 'center',
                overlayX: 'end',
                overlayY: 'center',
                offsetX: -this.offset,
            };
            switch (this.placement) {
                case 'top':
                    position = [positionTop, positionBottom];
                    break;
                case 'bottom':
                    position = [positionBottom, positionTop];
                    break;
                case 'left':
                    position = [positionLeft, positionRight, positionTop, positionBottom];
                    break;
                case 'right':
                    position = [positionRight, positionLeft, positionTop, positionBottom];
                    break;
                default:
                    break;
            }
            return position;
        };
        MdbPopoverDirective.prototype.show = function () {
            var _this = this;
            if (this._hideTimeout) {
                this._overlayRef.detach();
                clearTimeout(this._hideTimeout);
                this._hideTimeout = null;
            }
            this._createOverlay();
            if (this._hideTimeout) {
                clearTimeout(this._hideTimeout);
                this._hideTimeout = null;
            }
            this._showTimeout = setTimeout(function () {
                var tooltipPortal = new portal.ComponentPortal(MdbPopoverComponent);
                _this.popoverShow.emit(_this);
                _this._open = true;
                _this._tooltipRef = _this._overlayRef.attach(tooltipPortal);
                _this._tooltipRef.instance.content = _this.template || _this.mdbPopover;
                _this._tooltipRef.instance.title = _this.mdbPopoverTitle;
                _this._tooltipRef.instance.animation = _this.animation;
                _this._tooltipRef.instance.animationState = 'visible';
                _this._tooltipRef.instance.markForCheck();
                _this.popoverShown.emit(_this);
            }, this.delayShow);
        };
        MdbPopoverDirective.prototype.hide = function () {
            var _this = this;
            if (this._showTimeout) {
                clearTimeout(this._showTimeout);
                this._showTimeout = null;
            }
            else {
                return;
            }
            this._hideTimeout = setTimeout(function () {
                _this.popoverHide.emit(_this);
                if (!_this._tooltipRef) {
                    _this._overlayRef.detach();
                    _this._open = false;
                    _this.popoverHidden.emit(_this);
                }
                else {
                    _this._tooltipRef.instance._hidden.pipe(operators.first()).subscribe(function () {
                        _this._overlayRef.detach();
                        _this._open = false;
                        _this.popoverHidden.emit(_this);
                    });
                    _this._tooltipRef.instance.animationState = 'hidden';
                    _this._tooltipRef.instance.markForCheck();
                }
            }, this.delayHide);
        };
        MdbPopoverDirective.prototype.toggle = function () {
            if (this._open) {
                this.hide();
            }
            else {
                this.show();
            }
        };
        return MdbPopoverDirective;
    }());
    MdbPopoverDirective.decorators = [
        { type: core.Directive, args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: '[mdbPopover]',
                    exportAs: 'mdbPopover',
                },] }
    ];
    MdbPopoverDirective.ctorParameters = function () { return [
        { type: overlay.Overlay },
        { type: overlay.OverlayPositionBuilder },
        { type: core.ElementRef }
    ]; };
    MdbPopoverDirective.propDecorators = {
        mdbPopover: [{ type: core.Input }],
        mdbPopoverTitle: [{ type: core.Input }],
        popoverDisabled: [{ type: core.Input }],
        placement: [{ type: core.Input }],
        template: [{ type: core.Input }],
        animation: [{ type: core.Input }],
        trigger: [{ type: core.Input }],
        delayShow: [{ type: core.Input }],
        delayHide: [{ type: core.Input }],
        offset: [{ type: core.Input }],
        popoverShow: [{ type: core.Output }],
        popoverShown: [{ type: core.Output }],
        popoverHide: [{ type: core.Output }],
        popoverHidden: [{ type: core.Output }]
    };

    var MdbPopoverModule = /** @class */ (function () {
        function MdbPopoverModule() {
        }
        return MdbPopoverModule;
    }());
    MdbPopoverModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [common.CommonModule, overlay.OverlayModule],
                    declarations: [MdbPopoverDirective, MdbPopoverComponent],
                    exports: [MdbPopoverDirective, MdbPopoverComponent],
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.MdbPopoverDirective = MdbPopoverDirective;
    exports.MdbPopoverModule = MdbPopoverModule;
    exports.ɵa = MdbPopoverComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
