// mdb-angular-ui-kit/bundles/mdb-angular-ui-kit-tabs.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/cdk/coercion'), require('@angular/cdk/portal'), require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('mdb-angular-ui-kit/tabs', ['exports', '@angular/cdk/coercion', '@angular/cdk/portal', '@angular/core', 'rxjs', 'rxjs/operators', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['mdb-angular-ui-kit'] = global['mdb-angular-ui-kit'] || {}, global['mdb-angular-ui-kit'].tabs = {}), global.ng.cdk.coercion, global.ng.cdk.portal, global.ng.core, global.rxjs, global.rxjs.operators, global.ng.common));
}(this, (function (exports, coercion, portal, core, rxjs, operators, common) { 'use strict';

    var MDB_TAB_CONTENT = new core.InjectionToken('MdbTabContentDirective');
    var MdbTabContentDirective = /** @class */ (function () {
        function MdbTabContentDirective(template) {
            this.template = template;
        }
        return MdbTabContentDirective;
    }());
    MdbTabContentDirective.decorators = [
        { type: core.Directive, args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: '[mdbTabContent]',
                    providers: [{ provide: MDB_TAB_CONTENT, useExisting: MdbTabContentDirective }],
                },] }
    ];
    MdbTabContentDirective.ctorParameters = function () { return [
        { type: core.TemplateRef }
    ]; };

    var MDB_TAB_TITLE = new core.InjectionToken('MdbTabTitleDirective');
    var MdbTabTitleDirective = /** @class */ (function () {
        function MdbTabTitleDirective(template) {
            this.template = template;
        }
        return MdbTabTitleDirective;
    }());
    MdbTabTitleDirective.decorators = [
        { type: core.Directive, args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: '[mdbTabTitle]',
                    providers: [{ provide: MDB_TAB_TITLE, useExisting: MdbTabTitleDirective }],
                },] }
    ];
    MdbTabTitleDirective.ctorParameters = function () { return [
        { type: core.TemplateRef }
    ]; };

    var MdbTabComponent = /** @class */ (function () {
        function MdbTabComponent(_elementRef, _renderer, _vcr) {
            this._elementRef = _elementRef;
            this._renderer = _renderer;
            this._vcr = _vcr;
            this.activeStateChange$ = new rxjs.Subject();
            this._disabled = false;
            this._contentPortal = null;
            this._titlePortal = null;
            this._active = false;
        }
        Object.defineProperty(MdbTabComponent.prototype, "disabled", {
            get: function () {
                return this._disabled;
            },
            set: function (value) {
                this._disabled = coercion.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(MdbTabComponent.prototype, "active", {
            get: function () {
                return this._active;
            },
            // eslint-disable-next-line @typescript-eslint/adjacent-overload-signatures
            set: function (value) {
                if (value) {
                    this._renderer.addClass(this._elementRef.nativeElement, 'show');
                    this._renderer.addClass(this._elementRef.nativeElement, 'active');
                }
                else {
                    this._renderer.removeClass(this._elementRef.nativeElement, 'show');
                    this._renderer.removeClass(this._elementRef.nativeElement, 'active');
                }
                this._active = value;
                this.activeStateChange$.next(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(MdbTabComponent.prototype, "content", {
            get: function () {
                return this._contentPortal;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(MdbTabComponent.prototype, "titleContent", {
            get: function () {
                return this._titlePortal;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(MdbTabComponent.prototype, "shouldAttach", {
            get: function () {
                return this._lazyContent === undefined;
            },
            enumerable: false,
            configurable: true
        });
        MdbTabComponent.prototype.ngOnInit = function () {
            this._createContentPortal();
            if (this._titleContent) {
                this._createTitlePortal();
            }
        };
        MdbTabComponent.prototype._createContentPortal = function () {
            var content = this._lazyContent || this._content;
            this._contentPortal = new portal.TemplatePortal(content, this._vcr);
        };
        MdbTabComponent.prototype._createTitlePortal = function () {
            this._titlePortal = new portal.TemplatePortal(this._titleContent, this._vcr);
        };
        return MdbTabComponent;
    }());
    MdbTabComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'mdb-tab',
                    template: "<ng-template><ng-content></ng-content></ng-template>\n"
                },] }
    ];
    MdbTabComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.Renderer2 },
        { type: core.ViewContainerRef }
    ]; };
    MdbTabComponent.propDecorators = {
        _lazyContent: [{ type: core.ContentChild, args: [MDB_TAB_CONTENT, { read: core.TemplateRef, static: true },] }],
        _titleContent: [{ type: core.ContentChild, args: [MDB_TAB_TITLE, { read: core.TemplateRef, static: true },] }],
        _content: [{ type: core.ViewChild, args: [core.TemplateRef, { static: true },] }],
        disabled: [{ type: core.Input }],
        title: [{ type: core.Input }]
    };

    var MdbTabChange = /** @class */ (function () {
        function MdbTabChange() {
        }
        return MdbTabChange;
    }());
    var MdbTabsComponent = /** @class */ (function () {
        function MdbTabsComponent() {
            this._destroy$ = new rxjs.Subject();
            this._fill = false;
            this._justified = false;
            this._pills = false;
            this._vertical = false;
            this.activeTabChange = new core.EventEmitter();
        }
        Object.defineProperty(MdbTabsComponent.prototype, "fill", {
            get: function () {
                return this._fill;
            },
            set: function (value) {
                this._fill = coercion.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(MdbTabsComponent.prototype, "justified", {
            get: function () {
                return this._justified;
            },
            set: function (value) {
                this._justified = coercion.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(MdbTabsComponent.prototype, "pills", {
            get: function () {
                return this._pills;
            },
            set: function (value) {
                this._pills = coercion.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(MdbTabsComponent.prototype, "vertical", {
            get: function () {
                return this._vertical;
            },
            set: function (value) {
                this._vertical = coercion.coerceBooleanProperty(value);
            },
            enumerable: false,
            configurable: true
        });
        MdbTabsComponent.prototype.ngAfterContentInit = function () {
            var _this = this;
            var firstActiveTabIndex = this.tabs.toArray().findIndex(function (tab) { return !tab.disabled; });
            this.setActiveTab(firstActiveTabIndex);
            this.tabs.changes.pipe(operators.takeUntil(this._destroy$)).subscribe(function () {
                var hasActiveTab = _this.tabs.find(function (tab) { return tab.active; });
                if (!hasActiveTab) {
                    var closestTabIndex = _this._getClosestTabIndex(_this._selectedIndex);
                    if (closestTabIndex !== -1) {
                        _this.setActiveTab(closestTabIndex);
                    }
                }
            });
        };
        MdbTabsComponent.prototype.setActiveTab = function (index) {
            var activeTab = this.tabs.toArray()[index];
            if (!activeTab || (activeTab && activeTab.disabled)) {
                return;
            }
            this.tabs.forEach(function (tab) { return (tab.active = tab === activeTab); });
            this._selectedIndex = index;
            var tabChangeEvent = this._getTabChangeEvent(index, activeTab);
            this.activeTabChange.emit(tabChangeEvent);
        };
        MdbTabsComponent.prototype._getTabChangeEvent = function (index, tab) {
            var event = new MdbTabChange();
            event.index = index;
            event.tab = tab;
            return event;
        };
        MdbTabsComponent.prototype._getClosestTabIndex = function (index) {
            var tabs = this.tabs.toArray();
            var tabsLength = tabs.length;
            if (!tabsLength) {
                return -1;
            }
            for (var i = 1; i <= tabsLength; i += 1) {
                var prevIndex = index - i;
                var nextIndex = index + i;
                if (tabs[prevIndex] && !tabs[prevIndex].disabled) {
                    return prevIndex;
                }
                if (tabs[nextIndex] && !tabs[nextIndex].disabled) {
                    return nextIndex;
                }
            }
            return -1;
        };
        MdbTabsComponent.prototype.ngOnDestroy = function () {
            this._destroy$.next();
            this._destroy$.complete();
        };
        return MdbTabsComponent;
    }());
    MdbTabsComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'mdb-tabs',
                    template: "<ul\n  class=\"nav mb-3 col-3 flex-column\"\n  [ngClass]=\"{\n    'nav-pills': pills,\n    'nav-tabs': !pills,\n    'nav-fill': fill,\n    'nav-justified': justified,\n    'flex-column': vertical,\n    'col-3': vertical,\n    'text-center': vertical\n  }\"\n  role=\"tablist\"\n>\n  <li\n    *ngFor=\"let tab of tabs; let i = index\"\n    (click)=\"setActiveTab(i)\"\n    class=\"nav-item\"\n    role=\"presentation\"\n  >\n    <a\n      href=\"javascript:void(0)\"\n      class=\"nav-link\"\n      [class.active]=\"tab.active\"\n      [class.disabled]=\"tab.disabled\"\n      role=\"tab\"\n    >\n      <ng-template [ngIf]=\"tab.titleContent\">\n        <ng-template [cdkPortalOutlet]=\"tab.titleContent\"></ng-template>\n      </ng-template>\n\n      <ng-template [ngIf]=\"!tab.titleContent\">{{ tab.title }}</ng-template>\n    </a>\n  </li>\n</ul>\n\n<div\n  class=\"tab-content\"\n  [ngClass]=\"{\n    'col-9': vertical\n  }\"\n>\n  <!-- <ng-content select=\"mdb-tab\"></ng-content> -->\n  <ng-container *ngFor=\"let tab of tabs\">\n    <div\n      class=\"tab-pane fade\"\n      [ngClass]=\"{\n        show: tab.active,\n        active: tab.active\n      }\"\n    >\n      <ng-template mdbTabPortalOutlet [tab]=\"tab\"></ng-template>\n    </div>\n  </ng-container>\n</div>\n"
                },] }
    ];
    MdbTabsComponent.ctorParameters = function () { return []; };
    MdbTabsComponent.propDecorators = {
        tabs: [{ type: core.ContentChildren, args: [MdbTabComponent,] }],
        fill: [{ type: core.Input }],
        justified: [{ type: core.Input }],
        pills: [{ type: core.Input }],
        vertical: [{ type: core.HostBinding, args: ['class.row',] }, { type: core.Input }],
        activeTabChange: [{ type: core.Output }]
    };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    // eslint-disable-next-line @angular-eslint/directive-class-suffix
    var MdbTabPortalOutlet = /** @class */ (function (_super) {
        __extends(MdbTabPortalOutlet, _super);
        function MdbTabPortalOutlet(_cfr, _vcr, _document) {
            var _this = _super.call(this, _cfr, _vcr, _document) || this;
            _this._destroy$ = new rxjs.Subject();
            return _this;
        }
        MdbTabPortalOutlet.prototype.ngOnInit = function () {
            var _this = this;
            _super.prototype.ngOnInit.call(this);
            if ((this.tab.shouldAttach || this.tab.active) && !this.hasAttached()) {
                this.attach(this.tab.content);
            }
            else {
                this.tab.activeStateChange$.pipe(operators.takeUntil(this._destroy$)).subscribe(function (isActive) {
                    if (isActive && !_this.hasAttached()) {
                        _this.attach(_this.tab.content);
                    }
                });
            }
        };
        MdbTabPortalOutlet.prototype.ngOnDestroy = function () {
            this._destroy$.next();
            this._destroy$.complete();
            _super.prototype.ngOnDestroy.call(this);
        };
        return MdbTabPortalOutlet;
    }(portal.CdkPortalOutlet));
    MdbTabPortalOutlet.decorators = [
        { type: core.Directive, args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: '[mdbTabPortalOutlet]',
                },] }
    ];
    MdbTabPortalOutlet.ctorParameters = function () { return [
        { type: core.ComponentFactoryResolver },
        { type: core.ViewContainerRef },
        { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] }
    ]; };
    MdbTabPortalOutlet.propDecorators = {
        tab: [{ type: core.Input }]
    };

    var MdbTabsModule = /** @class */ (function () {
        function MdbTabsModule() {
        }
        return MdbTabsModule;
    }());
    MdbTabsModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [
                        MdbTabComponent,
                        MdbTabContentDirective,
                        MdbTabTitleDirective,
                        MdbTabPortalOutlet,
                        MdbTabsComponent,
                    ],
                    imports: [common.CommonModule, portal.PortalModule],
                    exports: [
                        MdbTabComponent,
                        MdbTabContentDirective,
                        MdbTabTitleDirective,
                        MdbTabPortalOutlet,
                        MdbTabsComponent,
                    ],
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.MdbTabComponent = MdbTabComponent;
    exports.MdbTabContentDirective = MdbTabContentDirective;
    exports.MdbTabTitleDirective = MdbTabTitleDirective;
    exports.MdbTabsComponent = MdbTabsComponent;
    exports.MdbTabsModule = MdbTabsModule;
    exports.ɵa = MDB_TAB_CONTENT;
    exports.ɵb = MDB_TAB_TITLE;
    exports.ɵc = MdbTabPortalOutlet;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
